faststreams/input_stream.nim

```diff
@@ -1,192 +1,298 @@
+  ,
+  getLenSync: proc (s: InputStream): int
+                   {.nimcall, gcsafe, raises: [IOError, Defect].} =
+    distance(s.head, s.bufferEnd)
+)
+
+template vtableAddr*(vtable: InputStreamVTable): ptr InputStreamVTable =
+  ## This is a simple work-around for the somewhat broken side
+  ## effects analysis of Nim - reading from global let variables
+  ## is considered a side-effect.
+  {.noSideEffect.}:
+    unsafeAddr vtable
+
+proc fileInput*(filename: string): InputStreamHandle =
+  let
+    memFile = memfiles.open(filename)
+    head = cast[ptr byte](memFile.mem)
+    fileSize = memFile.size
+
+  var stream = FileInputStream(
+    vtable: vtableAddr FileStreamVTable,
+    head: head,
+    bufferEnd: offset(head, fileSize),
+    bufferEndPos: fileSize,
+    file: memFile)
 
-proc openFile*(filename: string): ByteStreamVar =
-  new result
-  var memFile = memfiles.open(filename)
-  result.head = cast[ptr byte](memFile.mem)
   when debugHelpers:
-    result.bufferStart = result.head
-  result.bufferEnd = offset(result.head, memFile.size)
-  result.bufferEndPos = memFile.size
-  result.closeStream = proc = close memFile
-
-proc init*(T: type ByteStream,
-           mem: openarray[byte],
-           reader = StreamReader(nil),
-           asyncReader = AsyncStreamReader(nil)): ByteStream =
-  # TODO: the result should use `from mem` once it's supported
-  result.head = unsafeAddr mem[0]
-  result.bufferEnd = offset(result.head, mem.len)
-  result.bufferEndPos = mem.len
-  result.reader = reader
-  result.asyncReader = asyncReader
-
-proc memoryStream*(mem: openarray[byte]): ByteStreamVar = # TODO: mark the result with `from mem`
-  new result
-  result[] = ByteStream.init(mem)
-
-proc memoryStream*(str: string): ByteStreamVar = # TODO: mark the result with `from str`
-  new result
-  result[] = init(ByteStream, str.toOpenArrayByte(0, str.len - 1))
-
-proc init*(T: type BufferedStream,
-           reader = StreamReader(nil),
-           asyncReader = AsyncStreamReader(nil)): BufferedStream =
-  result.init result.buffer, reader, asyncReader
-
-proc endPos*(s: InputStreamVar): int =
-  # TODO This needs to use a VTable for `system.File` based streams
+    stream.bufferStart = head
+
+  InputStreamHandle(s: stream)
+
+proc memoryInput*(mem: openarray[byte]): InputStreamHandle =
+  let head = unsafeAddr mem[0]
+  InputStreamHandle(s: InputStream(
+    head: head,
+    bufferEnd: offset(head, mem.len),
+    bufferEndPos: mem.len))
+
+proc memoryInput*(str: string): InputStreamHandle =
+  memoryInput str.toOpenArrayByte(0, str.len - 1)
+
+# TODO: Is this used, should we deprecate it?
+proc endPos*(s: InputStream): int =
+  doAssert s.vtable == nil or s.vtable.getLenSync != nil
   return s.bufferEndPos
 
-proc syncRead(s: var ByteStream): bool =
-  let bytesRead = s.reader(s.bufferStart, s.bufferSize)
+# TODO The return type here could be Option[Natural] if Nim had
+# the Option[range] optimisation that will make it equvalent to `int`.
```

Comment:
I've had the same arguments in my head and might agree with them. For me, the primary motivation for using `Option` now is that this would allow for introducing the optimisation without breaking the API.

Reply:
> without breaking the API

yes, this is mostly how I'd prefer we reason about API and compiler improvements: ie if trivial performance or a missing optimization is all that stands between us and the better API, go with the better API unless the performance issue is truly significant - we can then motivate work on the compiler for these cases based on real-world usage. this is why using `Option` and `Result` at all makes sense - most of their perf issues are "solveable" without breaking code.